### DISCUSSION TITLE
Add possibility to switch to Bluetooth?

### DISCUSSION THREAD

**Discussion by natumbri** [CarlosDerSeher/snapclient#88] [Ideas]
Thanks for this port of snapclient - it's working great for me. I use it in conjunction with mopidy and the https://github.com/natumbri/mopidy-youtube project I look after.

One thing I've been thinking about - sometimes I like to switch from listening via my snapcast setup to listening to something directly from my phone, via Bluetooth. 

Would it be difficult to add the capacity to switch between snapcast and BT?

Something like this https://github.com/vedranMv/Ka-Radio32 but with snapclient instead of 'internet radio'.

I'd be happy to help, but am pretty clueless when it comes to writing real code!

Anyway, thanks again for the snapclient!

Cheers
Nik

**Comment by CarlosDerSeher** [comment]
Totally doable. We would need to add [this](https://github.com/espressif/esp-idf/tree/v5.2.1/examples%2Fbluetooth%2Fbluedroid%2Fclassic_bt%2Fa2dp_sink%2Fmain) as a new component and reuse pcm chunk memory allocation functions from snapclient. Only start advertising at first and if a connection is made all snapclient related task must be stopped, all memory freed and i2s handed over to bluetooth functions. If connection is dropped go back to snapclient mode. Easy enough 👌 😎 

I've always wanted to add bluetooth and stream from esp snapclient to snapserver and back again. Not sure if it is possible though.

I don't have much time to work on this project these days and my first priority is to resolve #54 and finish #69.

**Comment by DerPicknicker** [comment]
@CarlosDerSeher .. I think trying to run 2 instances of snapcast on one esp32 host would be also a Cool experiment ;-)

**Comment by CarlosDerSeher** [comment]
Why not just mount a second DAC of the same type?

**Comment by CarlosDerSeher** [comment]
2 instances would double the RAM and increase TCP bandwidth/RAM requirements

**Comment by DerPicknicker** [comment]
@CarlosDerSeher .. 2x the price and for example lyrat has PSRAM and also the boards from anabolyc.. this would be just a Cool experiment.

**Comment by CarlosDerSeher** [comment]
But where will you pipe the 2nd instance's output?

This is a little off topic here though, maybe you should open a discussion for this.

**Comment by ChristopherGriffin** [comment]
This would be great for some of my bluetooth portable beach speakers if I wanted them to sync with my home audio out by the pool.

Wifi>Bluetooth>speaker

**Comment by eni23** [comment]
I would be interested as well in a Bluetooth A2DP functionality. A nice feature would be to enable it over a GPIO pin instead of at boot time. While i would love to help implementing it, i doubt i have enough experience with esp-idf and c in general to be able to write good code, but maybe i can help sponsor it? 

Concerning connecting Bluetooth speakers to the esp and then streaming over Snapcast: i did this on Linux and had rather bad results, as there was a noteable delay. This could tho also be a result of Alsa&Pulseaudio and maybe it would work better on ESP, even tho WIFI&BT coexistence seems tricky to get right there.

**Reply by CarlosDerSeher** [reply]
Delay with bluetooth in the playback pipeline is normal and expected you won't be able to get this down to 0 as the speaker is responsible for the buffer allocation.

Regarding A2DP and sponsorship,  what do you have in mind exactly? This will for sure be a couple of nights worth of work.

**Reply by DerPicknicker** [reply]
Keep in mind that this will only work on esp32 and not on newer models like s3.

**Reply by CarlosDerSeher** [reply]
@DerPicknicker why?

**Reply by DerPicknicker** [reply]
> @DerPicknicker why?

Because esp32 removed the needed protocol: A2DP. Because this is part of the Bluetooth classic specification and the s3 board doesn't support it. Only BLE. 

If you search for ESP32-S3 Bluetooth audio you find more detailed information.

@eni23 .. Should it work outside local WiFi range? If not maybe just wait until MusicAssistant has this feature. 

If you're planning to build a esp32 powered Bluetooth speaker + snapcast it should run on the ESP32. 

WiFi and BL Coex: It's possible the only hard part is to find the right time to switch. You need to fill a buffer which is big enough to prevent audio stuttering. Because you're basically switching directly between BL and WiFi. You will definitely need a control mechanism if the buffer is empty and you can't make this too long because then you will notice stuttering. 

If you only want to use BL OR Snapcast it will be much easier. You don't need all of them.

**Reply by CarlosDerSeher** [reply]
Thank you I didn't realize that. Do you know if LE Audio is ready already?

**Reply by DerPicknicker** [reply]
@CarlosDerSeher no only H4. But H4 has no WiFi 😂
[See here](https://github.com/espressif/esp-idf/issues/12277#issuecomment-2249773871)

**Reply by CarlosDerSeher** [reply]
Yeah read I read this just now. A pitty they didn't implement it yet.

**Reply by BakersHub** [reply]
> I would be interested as well in a Bluetooth A2DP functionality. A nice feature would be to enable it over a GPIO pin instead of at boot time. While i would love to help implementing it, i doubt i have enough experience with esp-idf and c in general to be able to write good code, but maybe i can help sponsor it? 
> 
> Concerning connecting Bluetooth speakers to the esp and then streaming over Snapcast: i did this on Linux and had rather bad results, as there was a noteable delay. This could tho also be a result of Alsa&Pulseaudio and maybe it would work better on ESP, even tho WIFI&BT coexistence seems tricky to get right there.
> 
> 

I've got a version that operates into bluetooth a2dp at a press of a gpio if still interested in that style, I stopped working on it I favour for the following.

... 
No gpio needed
A a version that is snapcast and bluetooth both active 
But after bt connected and audio is sent, then snapcast is muted, wifi sent to low priority and bt operates as normal, 

When audio stopped from bt for 10 seconds/ or disconnected then snapcast unpauses and continues, includes qiuck ready to install .bin, you configure the rest on a ap hotspot, includes oled screen display and led controller sound reactive led driver, gpio volume control, ap hotspot enable via gpio, turn led lights on and off via gpio, fully configurable via device IP web ui 

On wrover only

**Comment by aximut** [comment]
I started working on a minimal implementation. No prior experience with A2DP but took a quick dive into it.
After some long debugging sessions, the implementation works mostly and supports connecting a device via Bluetooth and pre-defining a PIN for the connection (to keep the neighbors out ;P).
While connected, Snapcast keeps playing, until you start playing from your Bluetooth device, which will then override Snapcast. Once stopped, Snapcast will take over again.

The implementation is still very basic and needs some cleanup, more thread-safety mechanisms and optimization (in very rare occasions, it crashes on switch to Bluetooth and restarts, but no big deal as connection comes back). Overall it works quite nicely.
The good thing is that you can keep controlling the volume from Snapcast even while in Bluetooth mode.

Since I don't have the time to do this cleanup right now and open a proper PR, I will provide the patch that I did on top of commit `b484c2926b3fcbf3f0426094dfc8b3ab3494910b` and the corresponding changes to the `sdkconfig`. Happy if someone wants to take this over, the structure should be pretty clear. Meanwhile I'll keep enjoying Bluetooth and Snapcast music as it works exactly as I wanted it to work.

Have fun.

[snapclient_b484c2926b3fcbf3f0426094dfc8b3ab3494910b_bluetooth.patch](https://github.com/user-attachments/files/21565526/snapclient_b484c2926b3fcbf3f0426094dfc8b3ab3494910b_bluetooth.patch)
[sdkconfig.patch](https://github.com/user-attachments/files/21565850/sdkconfig.patch)

**Reply by CarlosDerSeher** [reply]
Looks quite nice and seems to be a good base for further improvement and integration to esp snapclient. Thanks for all the effort

**Reply by BakersHub** [reply]
i get error on line 188 trying to apply the patch

**Reply by BakersHub** [reply]
i got it working using vsc to manually apply the patch, now to add some features i discussed in https://github.com/sonocotta/esparagus-snapclient-fork/issues/25

mainly hardware volume control that works on both snap or bt, a led screen with time/eq/song details ect, and RGB lights that flash to beat ideally

everything i had with my 2 partition version that used gpio to switch between snap and bt

incorperated into this neater/simpler version


woud love some feedback on whether my twin partition is worth keeping/sharing to develop due to less radio inteference? or should this simpler version be better and forget the dual partition...

#wrover #a2dp #snapclient

**Reply by luar123** [reply]
I planned to look into bluetooth after improving inter task communication and adding player states (that would simplify coexistence). Probably only turning on BT while snapcast is not playing to reduce interference and hopefully make it work without psram.

**Reply by BakersHub** [reply]
awesome, sounds great, although you may have it worked out at a sweet spot as is.. will turning on bt only whilst not playing snapcast equal not broadcasting a BT connectivity? 

after testing yours im loving it, would of saved me alot of time and effort making my 2 parition version if i had found this earlier ... i started my esp journey in july and started the twin partition around september,  so suprised i never seen this lol

**Reply by luar123** [reply]
It is not mine ;-)

> will turning on bt only whilst not playing snapcast equal not broadcasting a BT connectivity?

I have not started to look into into it, but will need to test and measure sync to see what is feasable.

**Reply by luar123** [reply]
You all run on modules with PSRAM, right?

I looked into simply adding BT, no A2DP yet, just to see if it is feasible without PSRAM:

* playing snapcast + BT discoverable is not possible (BT takes ~80kB memory)
* disabling BT while playback still takes ~43kB memory. Could make it work with OPUS/570ms or PCM/680ms and no http_server. FLAC would require even less buffer. Usually I use flac/700ms (48kHz:16:2).
* freeing all BT memory while playback: Is an option, but requires an reboot after player stops.

There is definitely room for optimization, but I won't expect big improvements. And then the next question is if there is enough memory for a2dp playback when the player is stopped or if we need to deinit snapclient/wifi.

So at the moment I think it is possible to use it without psram, but there are trade-offs.

**Reply by BakersHub** [reply]
Yeah thats alot to ask for without psram

Have you tried this version on a wrover ??

https://github.com/BakersHub/Snapclient-BT-LED-OLED

Wrover


https://github.com/BakersHub/Snapclient-BT-LED-OLED/blob/main/ESPHOME%20WEB%20INSTALLER%20BINS%2FWROVERv1.2-latest.bin

Quickly install that on a wrover, reconfigure your pcm5102a or MAX98357A to your proffered gpios after install via AP or after wifi setup via device ip Web url

**Reply by luar123** [reply]
Thanks for sharing. I have only wroom modules here so I can't test your version. But I extracted the basic bluetooth functions and have it working here without PSRAM (with limitations). PR is almost ready.

**Reply by luar123** [reply]
See #217

**Comment by BakersHub** [comment]
https://github.com/BakersHub/Snapclient-BT-LED-OLED

Would love some feedback if anyone's got some wrovers laying around

Flashable bin via esphome, config via AP, 
If you have andoird phones laying around and would like to test the snapdroid apk I've tinkered with to work with the led controller setup

**Reply by aximut** [reply]
Thanks for looking into my patch and extending it. Looks very nice and useful but I don't have the time to test it right now

**Reply by CarlosDerSeher** [reply]
There is improv wifi available already for provisioning, is there a specific reason why you add the access point mode?

**Reply by BakersHub** [reply]
G'day Carlos! :),  it went this route to achieve just general access to a AP for anytime access to LED controller without a network needed

Some example use case
 - wifi network change, change wifi network without needing to reflash, hold ap goio for 3 seconds and get access to ap for password or ssid change or host ip change 
(I wanted this to be able to swap host IP on the fly from my home network, to my portable RPi4 snapserver which I take to remote places

- once configured, if wifi router was to fail, after 5 attempts of connection failed , it auto starts the AP: this may need adjusting to maybe 10 attempts so it's not false triggering, so far so good though...


 - standalone bluetooth speaker only with sound reactive led , if a user doesn't need use of snapcast, the device is still a very usefully bluetooth speaker with led controls 


So the most useful answer to your question would be the bluetooth ledcontroller when snapcast is not priority...

**Reply by CarlosDerSeher** [reply]
There is no need to reflash. I think in develop branch improv will be active for a few minutes after reboot. Still you'll need a serial connection which could be inconvinient sometimes I guess.

How do you think about creating a seperate component for bluetooth which we could pull in as a submodule? I am not a 100% sure how to integrate this but maybe providing some interface function  which registers a callback and if this returns true snapcast audio will be dropped in http task and then bluetooth task could own i2s functions. I a sure there will be more details to watch out for, but still it would be great to have this functionallity

**Reply by BakersHub** [reply]
yeah some of my small portable speaker builds may require no access to the serial usb without extensive disassembly of speaker or amp.
few builds have evolved around a broken UE BOOM or SONOS speakers where the original module has failed, and ive replaced the internals with the much superior Snapcast and dac and suited amp
so being able to remote install a small reset button is far more ideal is some cases, plus the added ability to add many configs to AP html page

Im 100% onboard, its defintely a feature thats needed from the ground up. to ensure a reliable bluetooth and wifi coexistence.

Im just very amature when it comes to this, im learning alot in the last 6 months, this being my first ever real attempt at coding in general, You guys would have way more of intelligence than me :P. im just a young boilermaker / welder / panel beater at the end of the day. LOL
I myself am relying on Ai for the most part but ill definetely have a go at making a component. 


<img width="705" height="379" alt="gtp" src="https://github.com/user-attachments/assets/379e1e88-8116-4243-a63f-b479ce23b2f4" />

**Reply by BakersHub** [reply]
<img width="768" height="345" alt="gtp" src="https://github.com/user-attachments/assets/fe3f8a6d-fc33-409f-84e9-5305d366c20b" />

**Reply by luar123** [reply]
Nice work
